File: config/claude/status_line.py
#!/usr/bin/env python3
"""Claude Code status line: minimal flat, monokai, starship-style git status."""

# pyright: basic

import json
import os
import subprocess
import sys

# Monokai
CYAN = "66d9ef"
YELLOW = "e6db74"
ORANGE = "fd971f"
RED = "f92672"
GRAY = "75715e"
LIGHT_GRAY = "a59f85"
LINE = "49483e"

RESET = "\033[0m"
BOLD = "\033[1m"

BRANCH_GLYPH = ""
BOLT = ""


def fg(hex_color: str) -> str:
    r = int(hex_color[0:2], 16)
    g = int(hex_color[2:4], 16)
    b = int(hex_color[4:6], 16)
    return f"\033[38;2;{r};{g};{b}m"


DIVIDER = f" {fg(LINE)}│{RESET} "


def get_git_info() -> tuple[str | None, str]:
    """Return (branch, starship-style status symbols) from one porcelain call."""
    try:
        result = subprocess.run(
            ["git", "--no-optional-locks", "status", "--porcelain=v2", "--branch"],
            capture_output=True,
            text=True,
            timeout=2,
        )
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        return (None, "")
    if result.returncode != 0:
        return (None, "")

    branch: str | None = None
    ahead = behind = 0
    conflicted = deleted = renamed = modified = staged = untracked = False

    for line in result.stdout.splitlines():
        if line.startswith("# branch.head "):
            head = line[len("# branch.head ") :]
            branch = None if head == "(detached)" else head
        elif line.startswith("# branch.ab "):
            parts = line.split()
            ahead = int(parts[2])
            behind = abs(int(parts[3]))
        elif line.startswith(("1 ", "2 ")):
            xy = line.split(" ", 2)[1]
            x, y = xy[0], xy[1]
            if x == "R" or y == "R":
                renamed = True
            if x == "D" or y == "D":
                deleted = True
            if x not in ".RD":
                staged = True
            if y not in ".RD":
                modified = True
        elif line.startswith("u "):
            conflicted = True
        elif line.startswith("? "):
            untracked = True

    symbols = ""
    if conflicted:
        symbols += "="
    if deleted:
        symbols += "✘"
    if renamed:
        symbols += "»"
    if modified:
        symbols += "!"
    if staged:
        symbols += "+"
    if untracked:
        symbols += "?"
    if ahead and behind:
        symbols += "⇕"
    elif ahead:
        symbols += "⇡"
    elif behind:
        symbols += "⇣"

    return (branch, symbols)


def env_flag_set(name: str) -> bool:
    """Return True when an env flag is set to a truthy value."""
    value = os.getenv(name)
    if not value:
        return False
    return value.strip().lower() not in {"0", "false"}


def get_runtime_badge() -> str | None:
    """Badge for the current runtime environment."""
    if env_flag_set("SANDBOX"):
        return "\U0001fae7"  # 🫧

    incus_container = os.getenv("INCUS_CONTAINER", "").strip()
    if incus_container:
        return f"\U0001f433 {fg(LIGHT_GRAY)}{incus_container}{RESET}"  # 🐳 name

    return None


def format_tokens(tokens: int) -> str:
    """Format token count nicely."""
    if tokens >= 1_000_000:
        return f"{tokens / 1_000_000:.1f}M"
    elif tokens >= 1000:
        return f"{tokens / 1000:.1f}k"
    return str(tokens)


def truncate(text: str, max_len: int = 20) -> str:
    """Truncate text with ellipsis."""
    if len(text) > max_len:
        return text[: max_len - 1] + "…"
    return text


def get_context_from_input(data: dict) -> tuple[int, float] | None:
    """Extract context info from Claude Code input (new API)."""
    if "context_window" in data:
        cw = data["context_window"]
        api_percentage = cw.get("used_percentage")
        window_size = cw.get("context_window_size", 200_000)
        if api_percentage is not None:
            tokens = int(window_size * api_percentage / 100)
            return (tokens, api_percentage)
    return None


def get_context_from_transcript(
    transcript_path: str, window_size: int = 200_000
) -> tuple[int, float] | None:
    """Fallback: parse transcript for context info."""
    try:
        with open(transcript_path) as f:
            lines = f.read().strip().split("\n")

        for line in reversed(lines):
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
                if (
                    obj.get("type") == "assistant"
                    and "message" in obj
                    and "usage" in obj["message"]
                    and "input_tokens" in obj["message"]["usage"]
                ):
                    usage = obj["message"]["usage"]
                    tokens = (
                        usage.get("input_tokens", 0)
                        + usage.get("cache_creation_input_tokens", 0)
                        + usage.get("cache_read_input_tokens", 0)
                    )
                    percentage = min(100, (tokens / window_size) * 100)
                    return (tokens, percentage)
            except (json.JSONDecodeError, KeyError):
                continue
    except OSError:
        pass
    return None


def main() -> None:
    data = json.load(sys.stdin)

    model = data.get("model", {}).get("display_name", "Claude")
    model = model.replace("Claude ", "")  # Shorten "Claude Opus 4.5" to "Opus 4.5"

    window_size = data.get("context_window", {}).get("context_window_size", 200_000)
    context_info = get_context_from_input(data)
    if not context_info and "transcript_path" in data:
        context_info = get_context_from_transcript(data["transcript_path"], window_size)

    parts = [f"{BOLD}{fg(CYAN)}{model}{RESET}"]

    if context_info:
        tokens, percentage = context_info
        # Percentage is against the auto-compact window when one is configured,
        # not the model's full context window.
        compact_window = os.getenv("CLAUDE_CODE_AUTO_COMPACT_WINDOW", "")
        if compact_window.isdigit() and int(compact_window) > 0:
            percentage = min(100, tokens / int(compact_window) * 100)
        text = f"{format_tokens(tokens)} {percentage:.0f}%"
        if percentage >= 90:
            parts.append(f"{BOLD}{fg(RED)}{BOLT} {text}{RESET}")
        elif percentage >= 80:
            parts.append(f"{fg(ORANGE)}{text}{RESET}")
        else:
            parts.append(f"{fg(GRAY)}{text}{RESET}")

    branch, symbols = get_git_info()
    if branch:
        git_part = f"{fg(YELLOW)}{BRANCH_GLYPH} {truncate(branch)}{RESET}"
        if symbols:
            git_part += f" {fg(ORANGE)}{symbols}{RESET}"
        parts.append(git_part)

    badge = get_runtime_badge()
    if badge:
        parts.append(badge)

    print(DIVIDER.join(parts))


if __name__ == "__main__":
    main()

File: config/pi/extensions/status-line.ts
import { execFile } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type {
	ExtensionAPI,
	ExtensionContext,
} from "@earendil-works/pi-coding-agent";
import { truncateToWidth, visibleWidth } from "@earendil-works/pi-tui";

const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM = "\x1b[2m";

const BRANCH_GLYPH = "";
const BOLT = "";

// Monokai
const MK = {
	cyan: "#66d9ef",
	yellow: "#e6db74",
	orange: "#fd971f",
	red: "#f92672",
	pink: "#f92672",
	purple: "#ae81ff",
	gray: "#75715e",
	lightGray: "#a59f85",
	line: "#49483e",
};

function hexToRgb(hex: string): [number, number, number] {
	const clean = hex.replace("#", "");
	return [
		Number.parseInt(clean.slice(0, 2), 16),
		Number.parseInt(clean.slice(2, 4), 16),
		Number.parseInt(clean.slice(4, 6), 16),
	];
}

function fgTrue(hex: string): string {
	const [r, g, b] = hexToRgb(hex);
	return `\x1b[38;2;${r};${g};${b}m`;
}

const DIVIDER = ` ${fgTrue(MK.line)}│${RESET} `;

/** Shorten known model ids, e.g. "gpt-5.6-sol" → "Sol 5.6". */
function modelDisplayName(id: string): string {
	const match = id.match(/gpt-(\d+(?:\.\d+)?)-(sol|terra|luna)/);
	if (match) {
		const [, version, codename] = match;
		return `${codename[0].toUpperCase()}${codename.slice(1)} ${version}`;
	}
	return id;
}

function formatTokens(tokens: number): string {
	if (tokens >= 1_000_000) return `${(tokens / 1_000_000).toFixed(1)}M`;
	if (tokens >= 1_000) return `${(tokens / 1_000).toFixed(1)}k`;
	return `${tokens}`;
}

function truncate(text: string, maxLen = 20): string {
	if (text.length <= maxLen) return text;
	return `${text.slice(0, maxLen - 1)}…`;
}

function stripAnsi(text: string): string {
	return text.replace(
		new RegExp(`${String.fromCharCode(27)}\\[[0-9;]*m`, "g"),
		"",
	);
}

export function normalizeExtensionStatuses(
	statuses: Iterable<string>,
): string[] {
	let hasFast = false;
	const normalized: string[] = [];

	for (const status of statuses) {
		const parts = status
			.split("•")
			.map((part) => part.trim())
			.filter((part) => part.length > 0);

		const keptParts = parts.filter((part) => {
			if (stripAnsi(part).trim().toLowerCase() !== "fast") return true;
			hasFast = true;
			return false;
		});

		if (keptParts.length > 0) {
			normalized.push(keptParts.join(" • "));
		}
	}

	return hasFast ? ["fast", ...normalized] : normalized;
}

export function partitionExtensionStatuses(statuses: Iterable<string>): {
	backgroundStatus?: string;
	remaining: string[];
} {
	let backgroundStatus: string | undefined;
	const remaining: string[] = [];
	for (const status of normalizeExtensionStatuses(statuses)) {
		const plain = stripAnsi(status).trim();
		if (/^(?:\s*\d+|(?:[●]\s*)?bg\s+\d+)$/i.test(plain)) {
			backgroundStatus = plain;
		} else {
			remaining.push(status);
		}
	}
	return { backgroundStatus, remaining };
}

function envFlagSet(name: string): boolean {
	const value = process.env[name];
	if (!value) return false;
	const normalized = value.trim().toLowerCase();
	return normalized !== "0" && normalized !== "false";
}

export function isCodexFastEnabled(
	configPath = join(homedir(), ".pi", "agent", "pi-codex-conversion.json"),
): boolean {
	try {
		if (!existsSync(configPath)) return false;
		const config = JSON.parse(readFileSync(configPath, "utf8")) as unknown;
		if (!config || typeof config !== "object") return false;
		const openai = (config as { openai?: unknown }).openai;
		return Boolean(
			openai &&
			typeof openai === "object" &&
			(openai as { fast?: unknown }).fast === true,
		);
	} catch {
		return false;
	}
}

function getRuntimeBadge(): string | null {
	if (envFlagSet("SANDBOX")) {
		return "🫧";
	}

	const incusContainer = process.env.INCUS_CONTAINER?.trim();
	if (incusContainer) {
		return `🐳 ${fgTrue(MK.lightGray)}${incusContainer}${RESET}`;
	}

	return null;
}

const GIT_POLL_INTERVAL_MS = 2_000;

export interface GitInfo {
	branch: string | null;
	symbols: string;
}

/** Parse `git status --porcelain=v2 --branch` into branch + starship-style symbols. */
export function parseGitStatusV2(out: string): GitInfo {
	let branch: string | null = null;
	let ahead = 0;
	let behind = 0;
	let conflicted = false;
	let deleted = false;
	let renamed = false;
	let modified = false;
	let staged = false;
	let untracked = false;

	for (const line of out.split("\n")) {
		if (line.startsWith("# branch.head ")) {
			const head = line.slice("# branch.head ".length);
			branch = head === "(detached)" ? "detached" : head;
		} else if (line.startsWith("# branch.ab ")) {
			const parts = line.split(" ");
			ahead = Math.abs(Number.parseInt(parts[2], 10)) || 0;
			behind = Math.abs(Number.parseInt(parts[3], 10)) || 0;
		} else if (line.startsWith("1 ") || line.startsWith("2 ")) {
			const xy = line.split(" ", 3)[1];
			const x = xy[0];
			const y = xy[1];
			if (x === "R" || y === "R") renamed = true;
			if (x === "D" || y === "D") deleted = true;
			if (x !== "." && x !== "R" && x !== "D") staged = true;
			if (y !== "." && y !== "R" && y !== "D") modified = true;
		} else if (line.startsWith("u ")) {
			conflicted = true;
		} else if (line.startsWith("? ")) {
			untracked = true;
		}
	}

	let symbols = "";
	if (conflicted) symbols += "=";
	if (deleted) symbols += "✘";
	if (renamed) symbols += "»";
	if (modified) symbols += "!";
	if (staged) symbols += "+";
	if (untracked) symbols += "?";
	if (ahead && behind) symbols += "⇕";
	else if (ahead) symbols += "⇡";
	else if (behind) symbols += "⇣";

	return { branch, symbols };
}

/**
 * Async branch + dirty-state poll. Complements Pi's git metadata watcher
 * (which only covers branch changes) and keeps git out of the render path.
 */
export function resolveGitInfo(cwd: string): Promise<GitInfo | null> {
	return new Promise((resolve) => {
		execFile(
			"git",
			["--no-optional-locks", "status", "--porcelain=v2", "--branch"],
			{ cwd, encoding: "utf8", timeout: 1_500, maxBuffer: 10 * 1024 * 1024 },
			(error, stdout) => {
				resolve(error ? null : parseGitStatusV2(stdout));
			},
		);
	});
}

function install(
	ctx: ExtensionContext,
	getCurrentCtx: () => ExtensionContext | undefined,
	getThinkingLevel: () => string,
) {
	ctx.ui.setFooter((tui, _theme, footerData) => {
		let branch = footerData.getGitBranch();
		let statusSymbols = "";
		let disposed = false;
		let gitPollTimer: ReturnType<typeof setTimeout> | undefined;

		const runGitPoll = () => {
			if (disposed) return;
			const cwd = getCurrentCtx()?.cwd ?? process.cwd();
			const startedAt = Date.now();
			void resolveGitInfo(cwd)
				.then((info) => {
					// null means git was unavailable or cwd is not a repository. Keep
					// the last known values rather than flickering on failures.
					if (disposed || !info) return;
					const nextBranch = info.branch ?? branch;
					if (nextBranch !== branch || info.symbols !== statusSymbols) {
						branch = nextBranch;
						statusSymbols = info.symbols;
						tui.requestRender();
					}
				})
				.finally(() => {
					if (disposed) return;
					// Back off proportionally in repos where status is slow, so
					// polling never uses more than a sliver of a core: a 20ms
					// status polls every 2s, a 1s status only every 20s.
					const duration = Date.now() - startedAt;
					gitPollTimer = setTimeout(
						runGitPoll,
						Math.max(GIT_POLL_INTERVAL_MS, duration * 20),
					);
					gitPollTimer.unref();
				});
		};

		const unsub = footerData.onBranchChange(() => {
			branch = footerData.getGitBranch();
			tui.requestRender();
		});
		runGitPoll();

		const safe = <T>(fn: () => T, fallback: T): T => {
			try {
				return fn();
			} catch {
				return fallback;
			}
		};

		return {
			dispose() {
				disposed = true;
				if (gitPollTimer) clearTimeout(gitPollTimer);
				unsub();
			},
			invalidate() {},
			render(width: number): string[] {
				const activeCtx = getCurrentCtx();
				const sessionName = activeCtx
					? safe(
							() => activeCtx.sessionManager.getSessionName()?.trim(),
							undefined,
						)
					: undefined;
				const runtimeBadge = getRuntimeBadge();
				const { backgroundStatus, remaining: extensionStatuses } =
					partitionExtensionStatuses(
						footerData.getExtensionStatuses().values(),
					);
				if (isCodexFastEnabled() && !extensionStatuses.includes("fast")) {
					extensionStatuses.unshift("fast");
				}
				const usage = activeCtx
					? safe(() => activeCtx.getContextUsage(), undefined)
					: undefined;

				let tokens = usage?.tokens ?? 0;
				let percent = usage?.percent ?? 0;
				if (!Number.isFinite(percent)) percent = 0;
				if (!Number.isFinite(tokens)) tokens = 0;

				const ctxText = `${formatTokens(Math.max(0, Math.round(tokens)))} ${percent.toFixed(0)}%`;
				let ctxPart: string;
				if (percent >= 90) {
					ctxPart = `${BOLD}${fgTrue(MK.red)}${BOLT} ${ctxText}${RESET}`;
				} else if (percent >= 80) {
					ctxPart = `${fgTrue(MK.orange)}${ctxText}${RESET}`;
				} else {
					ctxPart = `${fgTrue(MK.gray)}${ctxText}${RESET}`;
				}

				const model = activeCtx
					? safe(() => activeCtx.model, undefined)
					: undefined;
				const thinking = activeCtx ? safe(getThinkingLevel, "off") : "off";
				const modelPart = model
					? `${BOLD}${fgTrue(MK.cyan)}${modelDisplayName(model.id)}${RESET}`
					: `${fgTrue(MK.gray)}no-model${RESET}`;

				const parts: string[] = [modelPart, ctxPart];

				const cwd = activeCtx?.cwd ?? process.cwd();
				const dirname = cwd.replace(/\/+$/, "").split("/").pop() || "/";
				parts.push(`${fgTrue(MK.purple)}${truncate(dirname)}${RESET}`);

				if (branch) {
					let gitPart = `${fgTrue(MK.yellow)}${BRANCH_GLYPH} ${truncate(branch, 32)}${RESET}`;
					if (statusSymbols) {
						gitPart += ` ${fgTrue(MK.orange)}${statusSymbols}${RESET}`;
					}
					parts.push(gitPart);
				}

				if (runtimeBadge) {
					parts.push(runtimeBadge);
				}

				if (backgroundStatus) {
					parts.push(`${fgTrue(MK.pink)}${backgroundStatus}${RESET}`);
				}

				const left = parts.join(DIVIDER);

				const rightPieces: string[] = [];
				if (model?.reasoning) {
					rightPieces.push(thinking === "off" ? "thinking off" : thinking);
				}
				// Strip embedded ANSI so extension statuses can't override the
				// uniform dim styling of the right side.
				rightPieces.push(...extensionStatuses.map(stripAnsi));
				const right =
					rightPieces.length > 0
						? `${DIM}${rightPieces.join(" • ")}${RESET}`
						: "";

				const leftWidth = visibleWidth(left);
				const rightWidth = visibleWidth(right);
				const centerText = sessionName ? ` ${sessionName} ` : "";
				const centerSlot = width - leftWidth - rightWidth;
				if (centerText && centerSlot > 1) {
					const renderedCenter = truncateToWidth(centerText, centerSlot, "");
					if (renderedCenter) {
						const centerWidth = visibleWidth(renderedCenter);
						const centerPad = centerSlot - centerWidth;
						if (centerPad >= 0) {
							const leftPad = " ".repeat(Math.floor(centerPad / 2));
							const rightPad = " ".repeat(
								centerPad - Math.floor(centerPad / 2),
							);
							return [
								`${left}${leftPad}${DIM}${renderedCenter}${RESET}${rightPad}${right}`,
							];
						}
					}
				}

				if (leftWidth + 2 + rightWidth <= width) {
					const pad = " ".repeat(Math.max(2, width - leftWidth - rightWidth));
					return [`${left}${pad}${right}`];
				}

				return [truncateToWidth(left, width, "")];
			},
		};
	});
}

export default function (pi: ExtensionAPI) {
	let currentCtx: ExtensionContext | undefined;

	const apply = (ctx: ExtensionContext) => {
		currentCtx = ctx;
		install(
			ctx,
			() => currentCtx,
			() => pi.getThinkingLevel(),
		);
	};

	pi.on("session_start", async (_event, ctx) => {
		apply(ctx);
	});

	pi.on("session_shutdown", async (_event, ctx) => {
		currentCtx = undefined;
		ctx.ui.setFooter(undefined);
	});

	pi.registerCommand("statusline", {
		description: "Apply minimal monokai status footer",
		handler: async (_args, ctx) => {
			apply(ctx);
			ctx.ui.notify("Applied status line", "info");
		},
	});
}
